src/crypto/NoteEncryption.cpp

```diff
@@ -0,0 +1,164 @@
+#include "NoteEncryption.hpp"
+#include "crypto/sha512.h"
+#include <stdexcept>
+#include "sodium.h"
+#include <boost/static_assert.hpp>
+#include "prf.h"
+
+void clamp_curve25519(unsigned char *key)
+{
+    key[0] &= 248;
+    key[31] &= 127;
+    key[31] |= 64;
+}
+
+void KDF(unsigned char *K,
+    const uint256 &dhsecret,
+    const uint256 &epk,
+    const uint256 &pk_enc,
+    const uint256 &seed,
+    bool nonce_bit
+   )
+{
+    unsigned char K_tmp[64];
+    CSHA512 hasher;
+
+    // Trim the first byte from seed
+    std::vector<unsigned char> first_248(seed.begin() + 1, seed.end());
+
+    // First bit is the 'nonce' bit or `i`
+    // Second bit is zero
+    first_248[0] &= 0x3f;
+    first_248[0] |= (nonce_bit ? 1 : 0) << 7;
```

Comment:
For little-endian order:
```
first_248[0] &= 0xFC;
first_248[0] |= nonce_bit ? 1 : 0;
```

BTW, why is ``nonce_bit`` a bool? Of have thought it would be an integer that just happens to be constrained to 0 or 1.